File: output/rust/tests/11_state_var_reentry.rs
use std::collections::HashMap;

struct StateVarReentryFrameEvent {
    message: String,
    parameters: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl Clone for StateVarReentryFrameEvent {
    fn clone(&self) -> Self {
        Self {
            message: self.message.clone(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

impl StateVarReentryFrameEvent {
    fn new(message: &str) -> Self {
        Self {
            message: message.to_string(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

struct StateVarReentryFrameContext {
    event: StateVarReentryFrameEvent,
    _return: Option<Box<dyn std::any::Any>>,
    _data: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl StateVarReentryFrameContext {
    fn new(event: StateVarReentryFrameEvent, default_return: Option<Box<dyn std::any::Any>>) -> Self {
        Self {
            event,
            _return: default_return,
            _data: std::collections::HashMap::new(),
        }
    }
}

#[derive(Clone)]
struct StateVarReentryCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<StateVarReentryFrameEvent>,
}

impl StateVarReentryCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone, Default)]
struct CounterContext {
    count: i32,
}

#[derive(Clone)]
enum StateVarReentryStateContext {
    Counter(CounterContext),
    Other,
    Empty,
}

impl Default for StateVarReentryStateContext {
    fn default() -> Self {
        StateVarReentryStateContext::Counter(CounterContext::default())
    }
}

pub struct StateVarReentry {
    _state_stack: Vec<(String, StateVarReentryStateContext)>,
    __compartment: StateVarReentryCompartment,
    __next_compartment: Option<StateVarReentryCompartment>,
    _context_stack: Vec<StateVarReentryFrameContext>,
    _sv_count: i32,
}

impl StateVarReentry {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _context_stack: vec![],
            _sv_count: 0,
            __compartment: StateVarReentryCompartment::new("Counter"),
            __next_compartment: None,
        };
let __frame_event = StateVarReentryFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: StateVarReentryFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = StateVarReentryFrameEvent::new("<$");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = StateVarReentryFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = StateVarReentryFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &StateVarReentryFrameEvent) {
match self.__compartment.state.as_str() {
    "Counter" => self._state_Counter(__e),
    "Other" => self._state_Other(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: StateVarReentryCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Counter" => StateVarReentryStateContext::Counter(CounterContext { count: self._sv_count }),
    "Other" => StateVarReentryStateContext::Other,
    _ => StateVarReentryStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = StateVarReentryFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    StateVarReentryStateContext::Counter(ctx) => {
        self._sv_count = ctx.count;
    }
    StateVarReentryStateContext::Other => {}
    StateVarReentryStateContext::Empty => {}
}
    }

    pub fn increment(&mut self) -> i32 {
let mut __e = StateVarReentryFrameEvent::new("increment");
let __ctx = StateVarReentryFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Counter" => { self._s_Counter_increment(&__e); }
            "Other" => { self._s_Other_increment(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = StateVarReentryFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = StateVarReentryFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = StateVarReentryFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
let __ctx = self._context_stack.pop().unwrap();
if let Some(ret) = __ctx._return {
    *ret.downcast::<i32>().unwrap()
} else {
    Default::default()
}
    }

    pub fn get_count(&mut self) -> i32 {
let mut __e = StateVarReentryFrameEvent::new("get_count");
let __ctx = StateVarReentryFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Counter" => { self._s_Counter_get_count(&__e); }
            "Other" => { self._s_Other_get_count(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = StateVarReentryFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = StateVarReentryFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = StateVarReentryFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
let __ctx = self._context_stack.pop().unwrap();
if let Some(ret) = __ctx._return {
    *ret.downcast::<i32>().unwrap()
} else {
    Default::default()
}
    }

    pub fn go_other(&mut self) {
let mut __e = StateVarReentryFrameEvent::new("go_other");
let __ctx = StateVarReentryFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Counter" => { self._s_Counter_go_other(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = StateVarReentryFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = StateVarReentryFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = StateVarReentryFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
self._context_stack.pop();
    }

    pub fn come_back(&mut self) {
let mut __e = StateVarReentryFrameEvent::new("come_back");
let __ctx = StateVarReentryFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Other" => { self._s_Other_come_back(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = StateVarReentryFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = StateVarReentryFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = StateVarReentryFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
self._context_stack.pop();
    }

    fn _state_Other(&mut self, __e: &StateVarReentryFrameEvent) {
match __e.message.as_str() {
    "come_back" => { self._s_Other_come_back(__e); }
    "get_count" => { self._s_Other_get_count(__e); }
    "increment" => { self._s_Other_increment(__e); }
    _ => {}
}
    }

    fn _state_Counter(&mut self, __e: &StateVarReentryFrameEvent) {
match __e.message.as_str() {
    "get_count" => { self._s_Counter_get_count(__e); }
    "go_other" => { self._s_Counter_go_other(__e); }
    "increment" => { self._s_Counter_increment(__e); }
    "$>" => {
        self._sv_count = 0;
    }
    _ => {}
}
    }

    fn _s_Other_get_count(&mut self, __e: &StateVarReentryFrameEvent) {
-1;
    }

    fn _s_Other_increment(&mut self, __e: &StateVarReentryFrameEvent) {
-1;
    }

    fn _s_Other_come_back(&mut self, __e: &StateVarReentryFrameEvent) {
self.__transition(StateVarReentryCompartment::new("Counter"));
    }

    fn _s_Counter_go_other(&mut self, __e: &StateVarReentryFrameEvent) {
self.__transition(StateVarReentryCompartment::new("Other"));
    }

    fn _s_Counter_increment(&mut self, __e: &StateVarReentryFrameEvent) {
self._sv_count = self._sv_count + 1;
self._sv_count;
    }

    fn _s_Counter_get_count(&mut self, __e: &StateVarReentryFrameEvent) {
self._sv_count;
    }
}


fn main() {
    println!("=== Test 11: State Variable Reentry ===");
    let mut s = StateVarReentry::new();

    // Increment a few times
    s.increment();
    s.increment();
    let count = s.get_count();
    assert_eq!(count, 2, "Expected 2 after two increments, got {}", count);
    println!("Count before leaving: {}", count);

    // Leave the state
    s.go_other();
    println!("Transitioned to Other state");

    // Come back - state var should be reinitialized to 0
    s.come_back();
    let count = s.get_count();
    assert_eq!(count, 0, "Expected 0 after re-entering Counter (state var reinit), got {}", count);
    println!("Count after re-entering Counter: {}", count);

    // Increment again to verify it works
    let result = s.increment();
    assert_eq!(result, 1, "Expected 1 after increment, got {}", result);
    println!("After increment: {}", result);

    println!("PASS: State variables reinitialize on state reentry");
}

File: output/go/tests/11_state_var_reentry.go

package main

import "fmt"
import "os"

type StateVarReentryFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type StateVarReentryFrameContext struct {
    _event  StateVarReentryFrameEvent
    _return any
    _data   map[string]any
}

type StateVarReentryCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *StateVarReentryFrameEvent
    parentCompartment *StateVarReentryCompartment
}

func newStateVarReentryCompartment(state string) *StateVarReentryCompartment {
    return &StateVarReentryCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *StateVarReentryCompartment) copy() *StateVarReentryCompartment {
    nc := &StateVarReentryCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type StateVarReentry struct {
    _state_stack []*StateVarReentryCompartment
    __compartment *StateVarReentryCompartment
    __next_compartment *StateVarReentryCompartment
    _context_stack []StateVarReentryFrameContext
}

func NewStateVarReentry() *StateVarReentry {
    s := &StateVarReentry{}
    s._state_stack = make([]*StateVarReentryCompartment, 0)
    s._context_stack = make([]StateVarReentryFrameContext, 0)
    s.__compartment = newStateVarReentryCompartment("Counter")
    s.__next_compartment = nil
    __frame_event := StateVarReentryFrameEvent{_message: "$>", _parameters: nil}
    __ctx := StateVarReentryFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *StateVarReentry) __kernel(__e *StateVarReentryFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &StateVarReentryFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &StateVarReentryFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &StateVarReentryFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *StateVarReentry) __router(__e *StateVarReentryFrameEvent) {
    switch s.__compartment.state {
    case "Counter":
        s._state_Counter(__e)
    case "Other":
        s._state_Other(__e)
    }
}

func (s *StateVarReentry) __transition(next *StateVarReentryCompartment) {
    s.__next_compartment = next
}

func (s *StateVarReentry) Increment() int {
    __e := StateVarReentryFrameEvent{_message: "Increment"}
    __ctx := StateVarReentryFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *StateVarReentry) GetCount() int {
    __e := StateVarReentryFrameEvent{_message: "GetCount"}
    __ctx := StateVarReentryFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *StateVarReentry) GoOther() {
    __e := StateVarReentryFrameEvent{_message: "GoOther"}
    __ctx := StateVarReentryFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *StateVarReentry) ComeBack() {
    __e := StateVarReentryFrameEvent{_message: "ComeBack"}
    __ctx := StateVarReentryFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *StateVarReentry) _state_Other(__e *StateVarReentryFrameEvent) {
    if __e._message == "ComeBack" {
        __compartment := newStateVarReentryCompartment("Counter")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else if __e._message == "GetCount" {
        s._context_stack[len(s._context_stack)-1]._return = -1
        return
    } else if __e._message == "Increment" {
        s._context_stack[len(s._context_stack)-1]._return = -1
        return
    }
}

func (s *StateVarReentry) _state_Counter(__e *StateVarReentryFrameEvent) {
    __sv_comp := s.__compartment
    for __sv_comp != nil && __sv_comp.state != "Counter" { __sv_comp = __sv_comp.parentCompartment }
    if __e._message == "$>" {
        if _, ok := __sv_comp.stateVars["count"]; !ok {
            __sv_comp.stateVars["count"] = 0
        }
    } else if __e._message == "GetCount" {
        s._context_stack[len(s._context_stack)-1]._return = __sv_comp.stateVars["count"].(int)
        return
    } else if __e._message == "GoOther" {
        __compartment := newStateVarReentryCompartment("Other")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else if __e._message == "Increment" {
        __sv_comp.stateVars["count"] = __sv_comp.stateVars["count"].(int) + 1
        s._context_stack[len(s._context_stack)-1]._return = __sv_comp.stateVars["count"].(int)
        return
    }
}

func main() {
	fmt.Println("=== Test 11: State Variable Reentry ===")
	sm := NewStateVarReentry()

	// Increment a few times
	sm.Increment()
	sm.Increment()
	count := sm.GetCount()
	if count != 2 {
		fmt.Printf("FAIL: Expected 2 after two increments, got %d\n", count)
		os.Exit(1)
	}
	fmt.Printf("Count before leaving: %d\n", count)

	// Leave the state
	sm.GoOther()
	fmt.Println("Transitioned to Other state")

	// Come back - state var should be reinitialized to 0
	sm.ComeBack()
	count = sm.GetCount()
	if count != 0 {
		fmt.Printf("FAIL: Expected 0 after re-entering Counter (state var reinit), got %d\n", count)
		os.Exit(1)
	}
	fmt.Printf("Count after re-entering Counter: %d\n", count)

	// Increment again to verify it works
	result := sm.Increment()
	if result != 1 {
		fmt.Printf("FAIL: Expected 1 after increment, got %d\n", result)
		os.Exit(1)
	}
	fmt.Printf("After increment: %d\n", result)

	fmt.Println("PASS: State variables reinitialize on state reentry")
}
